core/client.go

```diff
@@ -121,118 +117,84 @@ func (c *Client) connect(ctx context.Context, addr string) error {
 		c.opts.Credential.Name(),
 		c.opts.Credential.Payload(),
 	)
-	err = c.WriteFrame(handshake)
-	if err != nil {
-		c.state = ConnStateRejected
+	if _, err := c.fs.WriteFrame(handshake); err != nil {
+		c.state = ConnStateDisconnected
 		return err
 	}
+
 	c.state = ConnStateConnected
 	c.localAddr = c.conn.LocalAddr().String()
 
 	c.logger.Printf("%s❤️  [%s][%s](%s) is connected to YoMo-Zipper %s", ClientLogPrefix, c.name, c.clientID, c.localAddr, addr)
 
 	// receiving frames
-	go c.handleFrame()
+	go func() {
+		closeConn, closeClient, err := c.handleFrame()
+		c.logger.Debugf("%shandleFrame: %v, %v, %T, %v", ClientLogPrefix, closeConn, closeClient, err, err)
+
+		c.mu.Lock()
+		defer c.mu.Unlock()
+
+		if c.state == ConnStateClosed {
+			return
+		}
+
+		c.state = ConnStateDisconnected
+		c.errc <- err
+
+		stream.Close()
+		if closeConn {
+			c.conn.CloseWithError(yerr.ErrorCodeClientAbort.To(), err.Error())
```

Comment:
Should more error type checking be added so that the server knows exactly what the error code is and why? For example, something like this:
```go
			if e, ok := err.(*quic.ApplicationError); ok {
				c.logger.Printf("%s[%s]connection closed by application error: %v", ClientLogPrefix, c.state, e)
				c.conn.CloseWithError(e.ErrorCode, e.Error())
			} else if e, ok := err.(*yerr.YomoError); ok {
				c.logger.Printf("%s[%s]connection closed by yomo error: %v", ClientLogPrefix, c.state, e)
				c.conn.CloseWithError(e.ErrorCode().To(), e.Error())
			} else {
				c.conn.CloseWithError(yerr.ErrorCodeClientAbort.To(), err.Error())
			}

```